### ISSUE TITLE
Robust standard error calculation for constraint estimations and better standard error interface

### ISSUE THREAD

**Issue by janosg** [OpenSourceEconomics/estimagic#57]
## Current Situation

- Standard error functions are relatively low level. The user explicitly calculates numerical derivatives and then passes them into the covariance/standard error functions
- It is possible (and happens frequently) that calculation of numerical derivatives fails because the stepsize is so large that some constraints are violated

## Desired Situation

- Standard error calculation can be done in one step
- Numerical derivatives should respect constraints

## Proposed Implementation

Write high level standard error functions that have a similar interface to maximize and minimize. Numerical derivatives are calculated on on a reparametrized version of the objective functions, such that all constraints hold. Then first calculate the covariance matrix for internal_params and use a parametric bootstrap (i.e. monte carlo integration) to calculate the covariance matrix of the external parameters

## Discussion

- Is anybody aware of literature on the topic? Is there any other approach that is similarly easy to apply?
- How would we handle simple box constraints?

**Comment by hmgaudecker** [comment]
- I do not think this is the way to go, see below. But if you do it anyhow, it seems just the quadratic version of #56, i.e., you multiply the each gradient in the product with the derivative of the constraint function. No bootstrap.
- At a more general level, I think this is adding too much complexity and scope for error. What do you do when you are on the border of the parameter space? 
- I would just pass the user's parameters and the constraints. If a constraint is violated, just go the other direction if doing finite differences (i.e. switch from central/forward to backward if upper bound constraint is violated); return `NaN` for the extrapolation (I would hope it can handle that, can't tell yet...). Sure this becomes more problematic for more general restrictions, but if you are calculating standard errors for a covariance matrix that becomes singular when adding &epsilon; you should not be calculating them in the first place.

**Comment by hmgaudecker** [comment]
Update: My calculation of the VCOV-matrix with Richardson's method finished and I would think that at some point a `NaN` was produced -- Jacobian looks good.

**Comment by janosg** [comment]
- It is the inverse of #56, but I thought we can do without cumbersome closed form solutions this time. But maybe you are right and we should do the work. 
- We would check boundary problems of course and refuse to calculate standard errors in that case.  
- Switch adaptively between forward central and backwards differences is hard to implement! We probably couldn't use numdifftools anymore. 
- If the derivative becomes undefined because we are close to a boundary (say a variance of size 1e-10) we should not calculate standard errors. But I was rather thinking of the case that for example an objective function can only be calculated if probabilities add up to one. Then the above approach would still work and I think standard errors would be valid. 

Maybe we postpone work on this issue until I talked to a statistician about this ...

**Comment by hmgaudecker** [comment]
> * It is the inverse of #56, but I thought we can do without cumbersome closed form solutions this time. But maybe you are right and we should do the work.
> 
> * We would check boundary problems of course and refuse to calculate standard errors in that case.

Then what is gained relative to the current situation?

> * Switch adaptively between forward central and backwards differences is hard to implement! We probably couldn't use numdifftools anymore.

Why? As long as the function returns `NaN` for invalid input as opposed to raising an error, I do not think you will need to worry much about numdifftools, see above. You could even ensure that via a wrapper, I have not thought through much whether that would seem like a good idea, though.

For the own implementation, you would loop over the derivatives: 
```python
try: 
    jac[i] = Derivative(func, x, method=method, bounds=bounds)
except ViolatedUpperBound as err:
    if method in {'central', 'forward'}:
        jac[i] = Derivative(func, x, method='backward', bounds=bounds)
    else:
        raise err, "something really weird happened here"
except ViolatedLowerBound as err:
    if method in {'central', 'backward'}:
        jac[i] = Derivative(func, x, method='backward', bounds=bounds)
    else:
        raise err, "something really weird happened here"
```

> * If the derivative becomes undefined because we are close to a boundary (say a variance of size 1e-10) we should not calculate standard errors. 

Fine. But this should not be hard-coded. For small enough values, a variance of 1e-10 might be perfectly fine... I would just try to make sure that the error message is clear to the user.

> But I was rather thinking of the case that for example an objective function can only be calculated if probabilities add up to one. Then the above approach would still work and I think standard errors would be valid.

That seems like an implementation detail. Say you have three categories for which the probabilities need to sum to one. You cannot calculate standard errors for all three anyhow. As in Multinomial Logit, the base category will have to adjust. If it is too small to accommodate the gradient step, you ~~have much worse problems than calculating standard errors~~ can always be a bit more fancy and check the adding up constraint, raising `Violated(Upper|Lower)Bound `as per above.

> 
> Maybe we postpone work on this issue until I talked to a statistician about this ...

**Comment by janosg** [comment]
Thanks for the code example. I hadn't thought about using scalar derivative functions from numdifftools as building blocks and just thought that we can't use their jacobian and hessian implementation directly anymore. Then this is really the way we should go.

For covariance constraints we could implement a closed form transformation (i.e. the inverse of #56)

Regarding the probability case: The bootstrap approach would yield standard errors for all three probabilities. Is this not a valid way?

**Comment by hmgaudecker** [comment]
> Thanks for the code example. I hadn't thought about using scalar derivative functions from numdifftools as building blocks and just thought that we can't use their jacobian and hessian implementation directly anymore. Then this is really the way we should go.

Good :-) I was a bit sloppy -- we will have to have a small wrapper that pretends we always have a function of one variable only.

> For covariance constraints we could implement a closed form transformation (i.e. the inverse of #56)

Ah, yes, and then we end up with the same type of violations. 

> Regarding the probability case: The bootstrap approach would yield standard errors for all three probabilities. Is this not a valid way?

There **are** only two independent probabilities = 2 parameters. The third is implicit via the adding up and non-negativity constraints. Same thing as why you only estimate one equation in a Probit model. 

If you allow the user to specify all three, you make it unnecessarily hard for the optimizer, I would think.

**Comment by janosg** [comment]
Of course there are only 2 free parameters, say p1 and p2 and the optimizer will only see (a transformed version of) those two. p3 = 1 - p1 - p2 is not a real parameter, but a test statistic, so we could calculate its standard error, either by simulation as I proposed originally or in closed form (`var(p3) = var(p1) + var(p2) + 2 cov(p1, p2)` or the matrix version of that for higher dimensions). 

The reason why I prefer to report all standard errors are:

- I want the user to fully abstract from the implementation of the constraints. The user only specifies 3 probabilities that sum up to one as start values, estimagic decides which of them becomes the base category and which transformation is applied to keep the probabilities between 0 and 1. 
- Some tests are easier if we have all standard errors. For example testing if p3 > 0 becomes trivial, otherwise I would have to do a test that that p1 + p2 < 1.
- It we can only calculate standard errors for a subset of the full parameter vector we need an api to decide which parameters are base categories and that is probably more work than filling the extra standard errors. 

I still don't propose to calculate standard errors for parameters that are explicitly fixed by the user, as in the case of a probit model or normalizations in skillmodels. But if a user decides to use the estimagic constraints for sum or probability constraints she probably wants to abstract from the fact that some of the parameters are not free.

**Comment by hmgaudecker** [comment]
Well, I find it much more intuitive to just specify two probabilities between zero and one rather than specifying three and making sure they sum to one. What about rounding errors for the latter? How does estimagic decide when I actually violate the constraint and when it is just reading/writing to disk  with too little precision? So Re your third point: I think it is a feature to just specify two parameters. 

Your users should be reasonably trained in statistics and be able to look up tests of the form you describe above in their favorite textbook. Else they have no business estimating complicated models in the first place. 

My prediction is that requiring people to specify an overidentified model in this way will be opening up cans of worms that we do not even see at this point.

**Comment by janosg** [comment]
I'm open to just specifying the free probabilities in the external parameter vector, even though I like that in the current implementation it is slightly easier to parse the parameter vector. 

The most important thing for me is that we get correct standard errors for all parameters in the external params vector that have not been explicitly fixed by the user. Not standard errors for the corresponding entry in the internal vector and not only for a subset of external parameters.

But I guess we can agree then on the following:

- We first calculate a covariance matrix on the internal parameter vector. For bound constraints we do that by choosing the stepsize and direction of the derivative accordingly. If a parameter is exactly at (or very close to) the boundary and therefore the standard errors are invalid warn the user. The transforming constraints (e.g. covariance matrices) are not a problem here.
- We then use closed form derivatives of the constraints and the delta method to get standard errors for all external parameters. If at all, a parametric bootstrap could be an optional alternative for this transformation, but for now we don't implement it. Whether all external parameters include the ones that are implicitly fixed is discussed in #61

**Comment by hmgaudecker** [comment]
Maybe we have been talking past each other here, but I would still suggest calculating things directly for the user's parameter vector. Much less hassle and maintenance burden for all normal cases. If things are close to the boundary, I am not sure one would gain a whole lot from working directly with the internal vector. Do you have an example where you would?

**Comment by janosg** [comment]
In respy we had the problem of invalid covariance matrices when using richardson extraploation. The outer step sizes can get quite large. But maybe this would have been solved by the rescaling you proposed in #59 and of course it would also be covered by the adaptive direction or step-size approach from your code snippet. 

The advantage of working with internal params in that case is probably only that we would need less trials and don't have the precision loss from choosing a non-optimal step-size or direction when the optimal ones fail. 

So for now we should solve #59, then implement the stuff from your code snippet and then check if we still have problems.

**Comment by hmgaudecker** [comment]
Cool.

Just curious, I thought the Richardson extrapolation would work if the largest step sizes returned NaN in the sense that numdifftools would then just use the smaller steps with valid results. Is that not the case?

**Comment by janosg** [comment]
This is quite possible but we did not catch the error and return a NaN. Thanks to @MaxBlesch I never really looked at numdifftools but that will change when I tackle #59

**Comment by hmgaudecker** [comment]
Since you understand the problem there well, it might be worth trying the current version in respy with a little wrapper (`return np.nan if some_constraint_violated(params_df) else func(params_df)`) to check. @tobiasraabe ?
